File: swift/Sources/Notes/Note.swift
import Foundation

class Note<Data: IdentifiableModel>: CanvasUnit {
    let storage: Storage<Data>
    let dateFormatter: DateFormatting
    private(set) var data: Data {
        didSet { storage.add(data) }
    }

    init(data: Data, storage: Storage<Data>, dateFormatter: DateFormatting) {
        self.storage = storage
        self.data = data
        self.dateFormatter = dateFormatter
        storage.add(data)
    }

    func update(_ newData: Data) {
        self.data = newData
    }

    func willRemove() {
        storage.remove(id: data.id)
    }

    func drawCanvas() -> String {
        "Note<\(String(describing: Data.self))> id=\(data.id)"
    }

    var id: String { data.id }
}

File: kotlin/src/main/kotlin/Note.kt
open class Note<Data : Identifiable>(
    data: Data,
    val storage: Storage<Data>,
    val dateFormatter: DateFormatting
) : CanvasUnit {

    var data: Data = data
        private set(value) {
            field = value
            storage.add(value)
        }

    init { storage.add(data) }

    open fun update(newData: Data) { this.data = newData }

    open fun willRemove() { storage.remove(data.id) }

    override fun drawCanvas(): String =
        "Note<${data::class.simpleName}> id=${data.id}"

    val id: String get() = data.id
}

interface AnyNote : CanvasUnit {
    val id: String
    fun willRemove()
}
